### PR TITLE
UnnecessaryFilter: add predicate functions

### LINKED ISSUE #6434: Missing predicate functions for `UnnecessaryFilter`
## Expected Behavior of the rule
The `UnnecessaryFilter`  rule should flag all terminal operations that have an equivalent that takes a predicate as parameter. 

```kotlin
// non-compliant
val x = listOf(1, 2, 3)
     .filter { it > 1 }
     .first()

// compliant
val x = listOf(1, 2, 3)
     .first { it > 1 }
```

The list should include at least: `any()`, `none()`,  `first()`, `firstOrNull()`, `last()`, `lastOrNull()`, `single()`, `singleOrNull()`.

### PR DESCRIPTION
**Pull request by t-kameyama**
Fixes #6434

### CODE REVIEW

**Review by BraisGabin — APPROVED** [review]
<3 really good one! Thanks!!

**Comment by BraisGabin** [comment]
Can you merge main again? There are conflicts with the previous PR 😅

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryFilterSpec.kt [hunk 1]
```diff
@@ -249,5 +323,16 @@ class UnnecessaryFilterSpec(val env: KotlinCoreEnvironment) {
             val findings = subject.compileAndLintWithContext(env, code)
             assertThat(findings).isEmpty()
         }
+
+        @Test
+        fun `Terminal function has an argument`() {
+            val code = """
+                val x = listOf(1, 2, 3).filter { it > 1 }.count { it > 2 }
+                val y = listOf(1, 2, 3).filter { it > 1 }.singleOrNull { it > 2 }
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryFilterSpec.kt:331]: 🤔 I'm not 100% sure about this. Should we allow these or should we flag them and force those conditions to be merged with an `&&`?

Comment (BraisGabin) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryFilterSpec.kt:331]: Then if someone opens an issue about this we can reconsider this. Meanwhile let's merge it.

Comment (t-kameyama) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryFilterSpec.kt:331]: https://github.com/detekt/detekt/actions/runs/6035265670/job/16375288840

https://github.com/detekt/detekt/blob/b5f40da57296d821239b7a4b54abb8ecdf582594/detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/RuleVisitor.kt#L146-L147

**Pull request merged**